### PR TITLE
Point selection interface

### PR DESCRIPTION
**Pull request by martukas**
This works, but I am not sure about my chosen function names. The general approach to storage seems ok, since there is no need for random access, though even that could be provided. I thought about vectors or arrays and other schemes, but there is probably nothing better out there. A bit concerned for the fact that it is legal to apply a selection of a mismatching rank to a dataspace... @eugenwintersberger  maybe you could look over this and give me some comments?
For sure needs some documentation before we merge.

### CODE REVIEW

File: src/h5cpp/dataspace/points.cpp [hunk 1]
```diff
@@ -26,35 +26,72 @@
...
-namespace dataspace {
+namespace hdf5
+{
+namespace dataspace
+{
 
 Points::Points() :
     Selection() {}
 
 Points::~Points() {}
```
Comment (matthew-d-jones) [inline src/h5cpp/dataspace/points.cpp]: Use `=default` and I suggest putting trivial constructors + destructors in the header file.

File: src/h5cpp/dataspace/points.hpp [hunk 2]
```diff
@@ -36,11 +36,22 @@ class DLL_EXPORT Points : public Selection {
...
   ~Points();
 
+  size_t rank() const;
+  size_t points() const;
+
+  void add(const std::vector<hsize_t>& coords);
+
+  void add_set(const std::vector<std::vector<hsize_t>>& coord_set);
+
   virtual void apply(const Dataspace &space,
```
Comment (matthew-d-jones) [inline src/h5cpp/dataspace/points.hpp]: Mark as `override` rather than `virtual`

File: src/h5cpp/dataspace/points.hpp [hunk 3]
```diff
@@ -36,11 +36,22 @@ class DLL_EXPORT Points : public Selection {
  public:
   Points();
   Points(size_t rank);
+  Points(const std::vector<std::vector<hsize_t>>& coord_set);
```
Comment (matthew-d-jones) [inline src/h5cpp/dataspace/points.hpp]: Both of these single arg constructors should be marked explicit.

File: test/dataspace/points_simple_test.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  dataspace::Simple space({10, 1024, 1024});
+  EXPECT_EQ(space.selection.type(), dataspace::SelectionType::ALL);
+
+  dataspace::Points points(2);
+  points.add({1, 1});
+  points.add({2, 2});
+  points.add({3, 3});
+  points.add({4, 4});
+
+  // TODO: why does this not throw when rank is not matching?
```
Comment (matthew-d-jones) [inline test/dataspace/points_simple_test.cpp]: `TODO` left in code, better to make a github issue.

File: test/node/dataset_partial_io_test.cpp [hunk 5]
```diff
@@ -65,6 +66,22 @@ TEST_F(PartialIO, test_read_write_scalar_int)
 }
 
 
+TEST_F(PartialIO, test_read_write_points)
+{
+  dataspace::Simple space{{10}};
+  auto type = datatype::create<int>();
+
+  node::Dataset dset(root_,Path("data"),type,space);
```
Comment (matthew-d-jones) [inline test/node/dataset_partial_io_test.cpp]: Spaces after commas are nice, also in the initialiser lists below.

File: test/node/dataset_partial_io_test.cpp [hunk 6]
```diff
@@ -65,6 +66,22 @@ TEST_F(PartialIO, test_read_write_scalar_int)
...
+
+  std::vector<int> write {1,2,3,4,5};
+  dset.write(write, dataspace::Points({{1}, {2}, {3}, {4}, {5}}));
```
Comment (matthew-d-jones) [inline test/node/dataset_partial_io_test.cpp]: Test is confusing, better to use values which don't match the indices they are written to.

File: test/node/dataset_partial_io_test.cpp [hunk 7]
```diff
@@ -65,6 +66,22 @@ TEST_F(PartialIO, test_read_write_scalar_int)
...
+
+  std::vector<int> read(3);
+  dset.read(read, dataspace::Points({{1}, {3}, {5}}));
+
+  EXPECT_EQ(read, std::vector<int>({1,3,5}));
+}
+
```
Comment (matthew-d-jones) [inline test/node/dataset_partial_io_test.cpp]: There's a lot of whitespace down here!

File: test/dataspace/points_simple_test.cpp (hunk 4 above)
Comment (martukas) [reply test/dataspace/points_simple_test.cpp]: I created a ticket, but I'm leaving this here so that I can see where the test needs to be changed.

**Pull request merged**